tracing/src/queries.rs

```diff
@@ -0,0 +1,264 @@
+use serde::Deserialize;
+use std::collections::{BTreeMap, HashSet};
+use tonic::codegen::tokio_stream::StreamExt;
+
+/// Runs a server that allows trace data to be queried and returned as a
+/// Firefox profile.
+pub async fn run_query_server(db: Database, port: u16) -> std::io::Result<()> {
+    HttpServer::new(move || {
+        App::new()
+            .wrap(Cors::permissive())
+            .app_data(web::Data::new(QueryState { db: db.clone() }))
+            .service(profile)
+    })
+    .bind(("0.0.0.0", port))?
+    .run()
+    .await
+}
+
+pub struct QueryState {
+    db: Database,
+}
+
+#[derive(Deserialize)]
+pub struct Query {
+    start_timestamp_unix_ms: i64,
+    end_timestamp_unix_ms: i64,
+    filter: QueryFilter,
+}
+
+#[derive(Deserialize)]
+pub struct QueryFilter {
+    nodes: HashSet<String>,
+    threads: HashSet<i64>, // TODO: filter by name, but the name doesn't exist nicely yet
+}
+
+pub struct QueryFilterResult {
+    matches: bool,
+    thread: i64,
+    node: String,
+}
+
+impl QueryFilter {
+    pub fn matches(&self, resource: &Resource, span: &Span) -> QueryFilterResult {
+        let mut node_name = "(unknown)".to_owned();
+        for attr in &resource.attributes {
+            if attr.key == "service.name" {
+                match &attr.value {
+                    Some(value) => match &value.value {
+                        Some(Value::StringValue(value)) => {
+                            node_name = value.clone();
+                        }
+                        _ => {}
+                    },
+                    _ => {}
+                }
+            }
+        }
+
+        let mut thread_id = -1;
+        for attr in &span.attributes {
+            if attr.key == "thread.id" {
+                match &attr.value {
+                    Some(value) => match &value.value {
+                        Some(Value::IntValue(value)) => {
+                            thread_id = *value;
+                        }
+                        _ => {}
+                    },
+                    _ => {}
+                }
+            }
+        }
+
+        let nodes_match = self.nodes.is_empty() || self.nodes.contains(&node_name);
+        let threads_match = self.threads.is_empty() || self.threads.contains(&thread_id);
+        QueryFilterResult {
+            matches: nodes_match && threads_match,
+            thread: thread_id,
+            node: node_name,
+        }
+    }
+}
+
+#[post("/profile")]
+async fn profile(
+    data: web::Data<QueryState>,
+    req: web::Json<Query>,
+) -> Result<HttpResponse, Error> {
+    // TODO: Enable compression as the data is very compressible
+    // TODO: Set a limit on the duration of the request interval.
+    let col = data.db.span_chunks();
+    let mut chunks = col
+        .find(
+            doc! {
+                "max_time": {"$gt": req.start_timestamp_unix_ms * 1000000},
+                "min_time": {"$lt": req.end_timestamp_unix_ms * 1000000},
+            },
```

Comment:
to confirm, here we are checking if there any intersection with the spanchunk right? and in this case, we still return the all the spans/events even if they may not be part of the intersection?

Reply:
Yep. We can add better filtering later.